### PR TITLE
Bug 1624626: Python: Add measure context manager

### PR DESCRIPTION
**Pull request by mdboom**

### CODE REVIEW

**Comment by badboy** [comment]
Looks like the wrong bug is mentioned here.

**Comment by badboy** [comment]
Fixed the PR title

File: glean-core/python/glean/metrics/timespan.py [hunk 1]
```diff
@@ -100,6 +100,37 @@ def cancel(self) -> None:
...
+        A context manager for recording timings. Used by the `measure` method.
+        """
+
+        def __init__(self, timespan: "TimespanMetricType"):
+            self._timespan = timespan
+
+        def __enter__(self) -> None:
+            self._timespan.start()
+
+        def __exit__(self, type, value, tb) -> None:
```
Comment (badboy) [inline glean-core/python/glean/metrics/timespan.py:114]: I read up on context managers and it seems all 3 values are referring to a potential exception.
Any reason to check tb specifically?
(According to https://docs.python.org/2/library/sys.html#sys.exc_info if there's no exception than all 3 are none, but it doesn't say that `tb` can't be `None` when there _is_ an exception. Then again I don't see how that would happen, so I defer that to you if this is correct)

File: glean-core/python/tests/metrics/test_timing_distribution.py [hunk 2]
```diff
@@ -150,3 +150,57 @@ def override_time_ns():
...
+        disabled=False,
+        category="telemetry",
+        lifetime=Lifetime.APPLICATION,
+        name=f"timing_distribution",
+        send_in_pings=["baseline"],
+        time_unit=TimeUnit.NANOSECOND,
+    )
+
+    with metric.measure():
+        override_time = 1000
```
Comment (badboy) [inline glean-core/python/tests/metrics/test_timing_distribution.py:176]: First reading that I was slightly confused. Maybe add a comment taht we actively manipulate the timing function?

File: glean-core/python/glean/metrics/timing_distribution.py [hunk 3]
```diff
@@ -142,6 +142,37 @@ def cancel():
...
+            else:
+                self._timing_distribution.cancel(self._timer_id)
+
+    def measure(self) -> "_TimingDistributionContextManager":
+        """
+        Provides a context manager for measuring the time it takes to execute
+        snippets of code in a `with` statement.
+
+        If the contents of the `with` statement raise an exception, the timing
+        is not recorded.
```
Comment (badboy) [inline glean-core/python/glean/metrics/timing_distribution.py:168]: Is the exception re-raised from that block?

File: glean-core/python/glean/metrics/timespan.py (hunk 1 above)
Comment (mdboom) [reply glean-core/python/glean/metrics/timespan.py:114]: I was just copying the code example in the PEP about context managers: https://docs.python.org/2.5/whatsnew/pep-343.html

But, yeah, maybe you're right that we need to be more careful.  This stuff isn't very well documented -- let me see if I can find a more canonical example.

File: glean-core/python/glean/metrics/timing_distribution.py (hunk 3 above)
Comment (mdboom) [reply glean-core/python/glean/metrics/timing_distribution.py:168]: Yes -- the `with` block doesn't catch the exception, so it bubbles out.

File: glean-core/python/glean/metrics/timespan.py (hunk 1 above)
Comment (badboy) [reply glean-core/python/glean/metrics/timespan.py:114]: That's good enough for me.

**Pull request merged**